File: App/Resources/AlgorithmDetails/flippedminheapsort/flippedminheapsort.py
def sort(arr):
    n = len(arr)

    def idx(p):
        return n - p

    def sift_down(root, dist):
        while root <= dist // 2:
            leaf = 2 * root
            if leaf < dist and arr[idx(leaf)] > arr[idx(leaf + 1)]:
                leaf += 1
            if arr[idx(root)] > arr[idx(leaf)]:
                arr[idx(root)], arr[idx(leaf)] = arr[idx(leaf)], arr[idx(root)]
                root = leaf
            else:
                break

    i = n // 2
    while i >= 1:
        sift_down(i, n)
        i -= 1

    i = n
    while i > 1:
        arr[idx(1)], arr[idx(i)] = arr[idx(i)], arr[idx(1)]
        sift_down(1, i - 1)
        i -= 1

array = [0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56]
sort(array)
print(array)

File: App/Resources/AlgorithmDetails/flippedminheapsort/flippedminheapsort.java
import java.util.Arrays;

public class flippedminheapsort {
  public static int idx(int p, int n) {
    return n - p;
  }

  public static void siftDown(int[] arr, int root, int dist, int n) {
    while (root <= dist / 2) {
      int leaf = 2 * root;
      if (leaf < dist && arr[idx(leaf, n)] > arr[idx(leaf + 1, n)]) {
        leaf += 1;
      }
      if (arr[idx(root, n)] > arr[idx(leaf, n)]) {
        int temp = arr[idx(root, n)];
        arr[idx(root, n)] = arr[idx(leaf, n)];
        arr[idx(leaf, n)] = temp;
        root = leaf;
      } else {
        break;
      }
    }
  }

  public static void sort(int arr[]) {
    int n = arr.length;

    int i = n / 2;
    while (i >= 1) {
      siftDown(arr, i, n, n);
      i -= 1;
    }

    i = n;
    while (i > 1) {
      int temp = arr[idx(1, n)];
      arr[idx(1, n)] = arr[idx(i, n)];
      arr[idx(i, n)] = temp;
      siftDown(arr, 1, i - 1, n);
      i -= 1;
    }
  }

  public static void main(String[] args) {
    int[] array = new int[]{0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};
    sort(array);
    System.out.println(Arrays.toString(array));
  }
}
